component/file_cache/file_cache.go

```diff
@@ -395,30 +399,48 @@ func (fc *FileCache) invalidateDirectory(name string) {
 		return
 	}
 	// TODO : wouldn't this cause a race condition? a thread might get the lock before we purge - and the file would be non-existent
+	var directoryStack []string
 	err = filepath.WalkDir(localPath, func(path string, d fs.DirEntry, err error) error {
 		if err == nil && d != nil {
-			log.Debug("FileCache::invalidateDirectory : %s (%d) getting removed from cache", path, d.IsDir())
+			// WalkDir goes through the tree in lexical order
+			// 	so 'aDir' always comes before 'aDir/file'
+			// 	and 'bDir' always comes after the whole 'aDir' subtree
+			// We can leverage this lexical order to queue folders for deletion after their children
+
+			// queue all files for deletion
 			if !d.IsDir() {
+				log.Debug("FileCache::invalidateDirectory : %s (file) getting removed from cache", path)
 				fc.policy.CachePurge(path)
-			} else {
-				deleteDirError := deleteFile(path)
-				if deleteDirError != nil {
-					log.Warn("FileCache::invalidateDirectory : %s - failed to delete %s. Here's why: %v", name, path, deleteDirError)
+			}
+			// queue folders for deletion after their children
+			if len(directoryStack) > 0 {
```

Comment:
Can you explain why you need to also do this here, when you also do a purge of everything at the end of the function?

Reply:
Technically I don't have to do this here. But it felt more satisfying to delete folders as soon as they're empty, and not all at the end.
But the code is more repetitive this way. I'll give you that.

For extra clarity, with our standard test nested folder structure:

- a/
  - a/c1/
    - a/c1/gc1
  - a/c2

this is the difference between this deletion sequence:

1. a/c1/gc1
2. a/c1/
3. a/c2
4. a

and this one:

1. a/c1/gc1
2. a/c2
3. a/c1/
4. a/

Reply:
Since it makes the code more repetitive, maybe you could just do it at the end, since it doesn't seem to improve performance. Also, since that the function is called for every file or directory, it seems inefficient to iterate over the list for each directory.

Reply:
I have just simplified this algorithm quite a bit. Since WalkDir is called in series, I just used it to build a list of items, which I then delete in reverse order. No more repetitive or complex logic. Just a bit more memory usage.